### DISCUSSION TITLE
Better organize components with optional reverse naming order

### DISCUSSION THREAD

**Discussion by DaniFoldi** [nuxt/nuxt#17622] [Ideas]
Hi,

I've been looking into creating some tree-shakable components for a project, however because of the folder-based naming structure, you can't organize components like this:
- `components/`
  - `Chart/`
    - `Bar.vue`
    - `Line.vue`

as they will be turned into `<ChartBar />` and `<ChartLine />` which looks silly in my opinion. I've found the relevant code that maps files to component names: [packages/nuxt/src/components/scan.ts](https://github.com/nuxt/framework/blob/109bd53037e38fb0322e4273f5452cc51521a0c8/packages/nuxt/src/components/scan.ts#L80)

and here is a patch to the current `dist/index.mjs` from line 538 that I have tested and it works nicely:
```js
while (prefixParts.length && (prefixParts[0] || "").toLowerCase() !== (fileNameParts[0] || "").toLowerCase()) {
        const part = prefixParts.shift();
        if (part.endsWith('+')) {
          componentNameParts.unshift(part);
        } else {
          componentNameParts.push(part);
        }
      }
      const componentName = fileName.endsWith('+') ? pascalCase(fileNameParts) + pascalCase(componentNameParts) : pascalCase(componentNameParts) + pascalCase(fileNameParts); // This may need a bit more thought
      const suffix = mode !== "all" ? `-${mode}` : "";
      if (resolvedNames.has(componentName + suffix) || resolvedNames.has(componentName)) {
        console.warn(`Two component files resolving to the same name \`${componentName}\`:

 - ${filePath}
 - ${resolvedNames.get(componentName)}`);
        continue;
      }
      resolvedNames.set(componentName + suffix, filePath);
      const pascalName = pascalCase(componentName).replace(/["'+]/g, "");
```
By renaming the files to `Bar+.vue` and `Line+.vue` I get `<BarChart />` and `<LineChart />` while keeping them inside `Chart/` - how lovely isn't it.

The source modification is quite similar, just different style. Happy to open a PR for this, happy to discuss alternatives, the functionality would be very convenient to have, even though I admit it might not be the easiest to understand with my implementation.

**Comment by pi0** [comment]
Would be nice idea to support it at least under a flag to change naming behavior. But what about 2+ nested dirs? `components/ui/common/Button` would be `ButtonCommonUI`?

**Reply by DaniFoldi** [reply]
(Haven't tested this thoroughly, so) **if** my implementation works like I imagined, `components/ui/common/Button` (exactly as you mentioned) would remain UICommonButton. In fact, let me create a table with all 8 possibilities:
(excluding components/)
|path|component|
|----|----------|
|`ui/common/Button`|`UICommonButton`|
|`ui/common/Button+`|`ButtonUICommon`|
|`ui/common+/Button`|`CommonUIButton`|
|`ui/common+/Button+`|`ButtonCommonUI`|
|`ui+/common/Button`|`UICommonButton`|
|`ui+/common/Button+`|`ButtonUICommon`|
|`ui+/common+/Button`|`CommonUIButton`|
|`ui+/common+/Button+`|`ButtonCommonUI`|

(Adding + to the first one doesn't change anything)
